modules/generator/registry/histogram.go

```diff
@@ -0,0 +1,206 @@
+	lastUpdated *atomic.Int64
+}
+
+var _ Histogram = (*histogram)(nil)
+var _ metric = (*histogram)(nil)
+
+func newHistogram(name string, labels []string, buckets []float64) *histogram {
+	bucketLabels := make([]string, len(buckets))
+	for i, bucket := range buckets {
+		bucketLabels[i] = formatFloat(bucket)
+	}
+
+	return &histogram{
+		nameCount:    fmt.Sprintf("%s_count", name),
+		nameSum:      fmt.Sprintf("%s_sum", name),
+		nameBucket:   fmt.Sprintf("%s_bucket", name),
+		labels:       labels,
+		buckets:      buckets,
+		bucketLabels: bucketLabels,
+		series:       make(map[uint64]*histogramSeries),
+	}
+}
+
+func (h *histogram) setCallbacks(onAddSeries func(count uint32) bool, onRemoveSeries func(count uint32)) {
+	h.onAddSerie = onAddSeries
+	h.onRemoveSerie = onRemoveSeries
+}
+
+func (h *histogram) Observe(labelValues *LabelValues, value float64) {
+	if len(h.labels) != len(labelValues.getValues()) {
+		panic(fmt.Sprintf("length of given label values does not match with labels, labels: %v, label values: %v", h.labels, labelValues))
+	}
+
+	hash := labelValues.getHash()
+
+	h.seriesMtx.RLock()
+	s, ok := h.series[hash]
+	h.seriesMtx.RUnlock()
+
+	if ok {
+		s.count.Add(1)
+		s.sum.Add(value)
+		for i, bucket := range h.buckets {
+			if value <= bucket {
+				s.buckets[i].Add(1)
+			}
+		}
+		s.bucketInf.Add(1)
+		s.lastUpdated.Store(time.Now().UnixMilli())
+		return
+	}
+
+	if !h.onAddSerie(h.activeSeriesPerHistogramSerie()) {
+		return
+	}
+
+	newSerie := &histogramSeries{
+		labelValues: labelValues.getValuesCopy(),
+		count:       atomic.NewFloat64(1),
+		sum:         atomic.NewFloat64(value),
+		buckets:     nil,
+		bucketInf:   atomic.NewFloat64(1),
+		lastUpdated: atomic.NewInt64(time.Now().UnixMilli()),
+	}
+	for _, bucket := range h.buckets {
+		if value <= bucket {
+			newSerie.buckets = append(newSerie.buckets, atomic.NewFloat64(1))
+		} else {
+			newSerie.buckets = append(newSerie.buckets, atomic.NewFloat64(0))
+		}
+	}
+
+	h.seriesMtx.Lock()
+	defer h.seriesMtx.Unlock()
+
+	s, ok = h.series[hash]
+	if ok {
+		s.count.Add(1)
+		s.sum.Add(value)
+		for i, bucket := range h.buckets {
+			if value <= bucket {
+				s.buckets[i].Add(1)
+			}
+		}
+		s.bucketInf.Add(1)
+		s.lastUpdated.Store(time.Now().UnixMilli())
+		return
+	}
+	h.series[hash] = newSerie
+}
+
+func (h *histogram) collectMetrics(appender storage.Appender, timeMs int64, externalLabels map[string]string) (activeSeries int, err error) {
+	h.seriesMtx.RLock()
+	defer h.seriesMtx.RUnlock()
+
+	activeSeries = len(h.series) * int(h.activeSeriesPerHistogramSerie())
+
+	lbls := make(labels.Labels, 1+len(externalLabels)+len(h.labels)+1)
+	lb := labels.NewBuilder(lbls)
+
+	// set external labels
+	for name, value := range externalLabels {
+		lb.Set(name, value)
+	}
+
+	for _, s := range h.series {
+		// set series-specific labels
+		for i, name := range h.labels {
+			lb.Set(name, s.labelValues[i])
+		}
+
+		// sum
+		lb.Set("__name__", h.nameSum)
```

Comment:
*nit*: This constant is exposed as `labels.MetricName`, since we are already using that for `le`.

Reply:
Done!